portfolio/src/main/java/com/google/sps/servlets/BusinessCardServlet.java

```diff
@@ -0,0 +1,173 @@
+    // Possible Errors:
+    // IllegalArgumentException - verify that entity was created properly
+    // ConcurrentModificationException - should not happen. Verify entity is not being edited.
+    // DatastoreFailureException - check Google Cloud Platform for more info.
+    Entity businessCardEntity = new Entity("BizCard");
+    businessCardEntity.setProperty("bizCard", bizCardBlobKey.getKeyString());
+    try {
+      datastore.put(businessCardEntity);
+    } catch (Exception e) {
+      response.sendError(500, e.getClass().getSimpleName() +
+          " There was an issue processing your upload.");
+      blobstoreService.delete(bizCardBlobKey);
+      return;
+    }
+
+    // Refresh the client page
+    response.sendRedirect("/index.html");
+  }
+
+  /**
+   * Return a list of blobKeys for all business cards in datastore.
+   * @param request the HTTP request sent from client for GET
+   * @param response HTTP response that will be sent back to the client
+   * @throws IOException if an IO error occurs while the request is being processed by the servlet.
+   */
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    // Create Query to get the blobKeys
+    Query query = new Query("BizCard");
+
+    // Get the Entities and populate the array list with the blobKeys
+    List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withDefaults());
+    List<String> blobKeys = results.stream().map(entity -> (String) entity.getProperty("bizCard"))
+        .collect(Collectors.toList());
+
+    // Convert list to JSON
+    String blobKeyJson = HttpServletUtilities.listToJson(blobKeys);
+
+    // Attach the JSON to the response
+    response.setContentType("application/json;");
+    response.getWriter().println(blobKeyJson);
+  }
+
+  /**
+   * This function will get the blobkey of an uploaded image from blobstore.
+   * Note this will only get the blobkey of the first image in the upload
+   * (If more than one file is uploaded per form element, this function will return
+   * unexpected results).
+   * @param request The HTTP request sent by blobstore
+   * @param paramName the name of the parameter in the request that stores the image
+   * @return blobkey of image
+   * @throws NullPointerException if no file was uploaded
+   * @throws IllegalArgumentException if the file was not an image
+   */
+  private BlobKey getImageBlobKey(HttpServletRequest request, String paramName)
+      throws FileNotFoundException, IllegalArgumentException {
+    // Get the relevant blobkeys from the blobstore
+    Map<String, List<BlobKey>> blobs = blobstoreService.getUploads(request);
+    List<BlobKey> blobKeys = blobs.get(paramName);
+
+    // Make sure at least one blob was uploaded
+    // In Dev, an empty file will be a null blob
+    // (thus this error will trigger in dev)
+    if (blobKeys == null || blobKeys.size() == 0) {
+      throw new FileNotFoundException();
```

Comment:
You are better off defining your own `InvalidBlob` exception type, and pass the desired message into its constructor. (In this case, "No images were uploaded. Please select an image to upload".) That way, you can handle all the exceptions with one `catch` statement and send the result of ToString() to the user.

This works because you are handling all the errors yourself, and you are using the same logic to handle them (except for the message).

Reply:
Cool idea - I never made my own Exception type before! 

Implemented - thanks!